drgn_tools/slabinfo.py

```diff
@@ -0,0 +1,310 @@
+        needswab = 1
+
+    if slab_freelist_hardened and needswab:
+        result = freelist_dereference_swab(prog, slab_cache, ptr_addr)
+    if not slab_freelist_hardened and not needswab:
+        try:
+            result = prog.read_word(ptr_addr)
+        except (AttributeError, drgn.ObjectAbsentError):
+            result = 0
+
+    return result
+
+
+def slub_get_freelist(
+    prog: drgn.Program,
+    slab_cache: drgn.Object,
+    freelist: drgn.Object,
+    freelist_set: Set[int],
+) -> None:
+    """
+    Main function to traverse through the freelist data of kmem_cache_cpu
+    """
+    ptr = freelist.value_()
+    while ptr:
+        if ptr & PAGE_MAPPING_ANON:
+            break
+        freelist_set.add(ptr)
+        ptr = try_hardened_freelist_dereference(
+            prog, slab_cache, (ptr + slab_cache.offset.value_())
+        )
+
+
+def slub_get_cpu_freelist(
+    prog: drgn.Program, slab_cache: drgn.Object, cpu_freelist: drgn.Object
+) -> set:
+    """
+    Collect freelist data from kmem_cache_cpu and
+    return the set of free objects
+    """
+    cpu_free_set: Set[int] = set()
+    slub_get_freelist(prog, slab_cache, cpu_freelist, cpu_free_set)
+    return cpu_free_set
+
+
+def slub_get_cpu_freelist_cnt(
+    prog: drgn.Program, slab_cache: drgn.Object, cpu_freelist: drgn.Object
+) -> int:
+    """
+    Collect freelist data from kmem_cache_cpu and
+    return the count of free objects
+    """
+    cpu_free_set_rc = slub_get_cpu_freelist(prog, slab_cache, cpu_freelist)
+    return len(cpu_free_set_rc)
+
+
+def slub_partial_free(
+    prog: drgn.Program, slab_cache: drgn.Object, slab: drgn.Object
+) -> int:
+    """
+    Collect partial list from kmem_cache_cpu
+    """
+    cpu_partial = slab
+    pss_cnt = 0
+    partial_free = partial_objects = partial_inuse = 0
+
+    while cpu_partial:
+        page = cast("struct page *", cpu_partial)
+
+        try:
+            partial_objects = page.objects.value_()
+        except (AttributeError, drgn.ObjectAbsentError):
+            partial_objects = 0
+            return 0
+
+        try:
+            partial_inuse = page.inuse.value_()
+        except (AttributeError, drgn.ObjectAbsentError):
+            partial_inuse = 0
+            return 0
+
+        partial_free += partial_objects - partial_inuse
+
+        try:
+            cpu_partial = page.next
+        except (AttributeError, drgn.ObjectAbsentError):
+            cpu_partial = 0
+        pss_cnt += 1
+
+    return partial_free
+
+
+def kmem_cache_slub_info(
+    prog: drgn.Program, slab: drgn.Object, nrnodeid: int
+) -> None:
+    """
+    Collect the per-slab data.
+    Traverse though all the CPUs to collect the slab data.
+    """
+    total_slabs = objects = free_objects = total_inuse = 0
+
+    for cpuid in for_each_online_cpu(prog):
+        try:
+            per_cpu_slab = per_cpu_ptr(slab.cpu_slab, cpuid)
+        except (AttributeError, drgn.ObjectAbsentError):
+            continue
```

Comment:
I'm commenting here, but I'm curious about it at pretty much every use: in what cases do you encounter `ObjectAbsentError`? As far as I understand, this only occurs when reading values out of stack traces.

As for `AttributeError`, I can see that would happen if the type `kmem_cache` didn't contain a `cpu_slab` field - which is is the case for systems where slab is enabled rather than slub. While we don't need to _support_ slab, we should at least raise an understandable error, rather than silently ignoring it and crashing later.

I guess what I'm driving at is, why are we catching these exceptions? Without a documented explanation, I don't really like having exception handlers. They tend to end up suppressing errors we would want to see, and if that happens, the resulting behavior is usually incomplete in some way that would harm our bug analysis. Having the program crash is really unfortunate, and I don't like that either, but it would be better to do that than have flawed debugging analyses.

Reply:
+1